library/std/src/sys/platform_version/darwin/mod.rs

```diff
@@ -0,0 +1,446 @@
+        }
+    }
+
+    // Introduced in macOS 10.13.4.
+    // https://github.com/apple-oss-distributions/xnu/blob/xnu-11215.81.4/bsd/kern/kern_sysctl.c#L2015-L2051
+    sysctl_version(c"kern.osproductversion")
+}
+
+/// Look up the current OS version(s) from `/System/Library/CoreServices/SystemVersion.plist`.
+///
+/// More specifically, from the `ProductVersion` and `iOSSupportVersion` keys, and from
+/// `$IPHONE_SIMULATOR_ROOT/System/Library/CoreServices/SystemVersion.plist` on the simulator.
+///
+/// This file was introduced in macOS 10.3, which is well below the minimum supported version by
+/// `rustc`, which is (at the time of writing) macOS 10.12.
+///
+/// # Implementation
+///
+/// We do roughly the same thing in here as `compiler-rt`, and dynamically look up CoreFoundation
+/// utilities for parsing PLists (to avoid having to re-implement that in here, as pulling in a full
+/// PList parser into `std` seems costly).
+///
+/// If this is found to be undesirable, we _could_ possibly hack it by parsing the PList manually
+/// (it seems to use the plain-text "xml1" encoding/format in all versions), but that seems brittle.
+fn version_from_plist() -> OSVersion {
+    // The root directory relative to where all files are located.
+    let root = if cfg!(target_abi = "sim") {
+        PathBuf::from(env::var_os("IPHONE_SIMULATOR_ROOT").expect(
+            "environment variable `IPHONE_SIMULATOR_ROOT` must be set when executing under simulator",
+        ))
+    } else {
+        PathBuf::from("/")
+    };
+
+    // Read `SystemVersion.plist`. Always present on Apple platforms, reading it cannot fail.
+    let path = root.join("System/Library/CoreServices/SystemVersion.plist");
+    let plist_buffer = fs::read(&path).unwrap_or_else(|e| panic!("failed reading {path:?}: {e}"));
+    parse_version_from_plist(&root, &plist_buffer)
+}
+
+/// Split out from [`version_from_plist`] to allow for testing.
+#[allow(non_upper_case_globals, non_snake_case)]
+fn parse_version_from_plist(root: &Path, plist_buffer: &[u8]) -> OSVersion {
+    // Link to the CoreFoundation dylib, and look up symbols from that.
+    // We explicitly use non-versioned path here, to allow this to work on older iOS devices.
+    let cf_path = root.join("System/Library/Frameworks/CoreFoundation.framework/CoreFoundation");
+
+    let cf_handle = run_path_with_cstr(&cf_path, &|path| unsafe {
+        Ok(libc::dlopen(path.as_ptr(), libc::RTLD_LAZY | libc::RTLD_LOCAL))
+    })
+    .expect("failed allocating string");
+    if cf_handle.is_null() {
+        let err = unsafe { CStr::from_ptr(libc::dlerror()) };
+        panic!("could not open CoreFoundation.framework: {err:?}");
+    }
+    let _cf_handle_free = Deferred(|| {
+        // Ignore errors when closing. This is also what `libloading` does:
+        // https://docs.rs/libloading/0.8.6/src/libloading/os/unix/mod.rs.html#374
+        let _ = unsafe { libc::dlclose(cf_handle) };
+    });
+
+    macro_rules! dlsym {
```

Comment:
Is it possible to reuse the logic at https://github.com/rust-lang/rust/blob/8a407a82848bbc926de1cbbbbcb381e1a96f5968/library/std/src/sys/pal/unix/weak.rs#L2?